src/Subscription.js

```diff
@@ -89,14 +74,72 @@ export default class Subscription extends EventEmitter {
             previousMsgRef.compareTo(this.lastReceivedMsgRef[key]) === 1
     }
 
-    handleMessage(msg, isResend = false) {
+    handleBroadcastMessage(msg, verificationPromise) {
+        return this._handleMessage(msg, verificationPromise, false)
+    }
+
+    handleResentMessage(msg, verificationPromise) {
+        if (!this.resending) {
+            throw new Error(`There is no resend in progress, but received resent message ${msg.serialize()}`)
+        } else {
+            const handleMessagePromise = this._handleMessage(msg, verificationPromise, true)
+            this._lastMessageHandlerPromise = handleMessagePromise
+            return handleMessagePromise
+        }
+    }
+
+    handleResending(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseResending message ${response.serialize()}`)
+        }
+        this.emit('resending', response)
+    }
+
+    handleResent(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseResent message ${response.serialize()}`)
+        }
+        // Delay event emission until the last message in the resend has been handled
+        if (this._lastMessageHandlerPromise) {
+            this._lastMessageHandlerPromise.then(() => {
+                this.emit('resent', response)
+                this._finishResend()
+            })
+        } else {
+            throw new Error('Attempting to handle ResendResponseResent, but no messages have been received!')
+        }
+    }
+
+    handleNoResend(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseNoResend message ${response.serialize()}`)
```

Comment:
We don't actually need to call `_finishResend` here, because the flag shows that no resend is in progress and thus there is nothing to clean up. 

This is a sanity check, which would basically occur if the socket spontaneously sent a phantom resend response without any such request being sent. It should never occur unless something's broken either in client code or server code. 

I added these assertions because they would have caught the issue fixed in this PR earlier, and it would not have creeped in undetected as a side effect of the signature verification feature.